fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/clientlog/KeySpaceCountTree.java

```diff
@@ -503,23 +509,42 @@ protected TupleKeyCountTree newPrefixChild(@Nonnull Object prefix) {
             }
         }
         if (recordTypeKeyspace != null) {
-            KeyExpression primaryKey = recordTypeKeyspace.getRecordType().getPrimaryKey();
-            if (distance + 1 < primaryKey.getColumnSize()) {
-                return resolvePrimaryKeyField(context, resolvedParent, object, primaryKey, distance + 1);
+            List<KeyExpression> storedPrimaryKeys = recordTypeKeyspace.getRecordType().getPrimaryKey().normalizeKeyForPositions();
+            if (distance + 1 < storedPrimaryKeys.size()) {
+                return resolvePrimaryKeyField(context, resolvedParent, object, storedPrimaryKeys.get(distance + 1), distance + 1);
             }
         }
         if (indexKeyspace != null &&
                 indexKeyspace.getParent() instanceof ResolvedRecordStoreKeyspace &&
                 ((ResolvedRecordStoreKeyspace)indexKeyspace.getParent()).getRecordStoreKeyspace() == FDBRecordStoreKeyspace.INDEX) {
             Index index = indexKeyspace.getIndex();
-            KeyExpression indexExpression = index.getRootExpression();
-            if (distance < indexExpression.getColumnSize()) {
-                return resolveIndexField(context, resolvedParent, object, index, indexExpression, distance);
+            List<KeyExpression> storedKeys = indexStoredKeys(index);
+            if (distance < storedKeys.size()) {
+                return resolveIndexField(context, resolvedParent, object, index, storedKeys.get(distance), distance);
             }
         }
         return UNRESOLVED;
     }
 
+    // TODO: Get this from the IndexMaintainerFactory via some new interface (the IndexMaintainer needs a RecordStore).
```

Comment:
I'm not opposed to trying, but as #854 points out, this might be more clear as something the new planner develops.

Reply:
I actually think we don't want to add this to the index maintainer yet, given (a) the new planner might change things/make things better and (b) this API is still experimental, so adding to the index maintainer factories now seems maybe a little premature.

I could see an argument, though, this method should be `protected` in case someone wants to provide an implementation for a custom index of their own devising by overriding this method.

Reply:
Yeah, that seems like a reasonable compromise.